docs/assessment.md

```diff
@@ -224,6 +224,122 @@ When calculating a student's grade for a group assessment, PrairieLearn will alw
+
+- Submitting specific questions
+- Viewing specific questions
+- Assigning group roles for other students
+
+Although in most cases each student is expected to take one role, students are allowed to take on multiple roles in some narrow scenarios, such as when users leave and join groups after assessments have started.
+
+To opt-in to custom group roles, group roles must be defined at the root of the `infoAssessment.json` file. For example:
+
+```json
+{
+  "groupRoles": [
+    {
+      "name": "Manager",
+      "minimum": 1,
+      "maximum": 1,
+      "canAssignRoles": true
+    },
+    {
+      "name": "Recorder",
+      "minimum": 1,
+      "maximum": 1
+    },
+    {
+      "name": "Reflector",
+      "minimum": 1,
+      "maximum": 1
+    },
+    {
+      "name": "Contributor"
+    }
+  ]
+}
+```
+
+| Attribute        | Type    | Default | Description                                                                  |
+| ---------------- | ------- | ------- | ---------------------------------------------------------------------------- |
+| `name`           | string  | -       | The name of the role.                                                        |
+| `minimum`        | integer | 0       | The minimum required number of students holding this role in the assessment. |
+| `maximum`        | integer | -       | The maximum required number of students holding this role in the assessment. |
+| `canAssignRoles` | boolean | false   | Allow students with this role to assign roles to other students.             |
+
+Students typically select their roles before starting an assessment, but they can change their roles mid-assessment if needed. As a safeguard against invalid role configurations, PrairieLearn prevents students from viewing questions if a group's role configuration does not meet the instructor's specification.
+
+#### Adding permissions for an assessment
+
+Permissions can be configured at the _assessment_, _zone_, or _question_ level.
+
+The schema for permissions is defined as follows:
+
+```json
+{
+  "canView": ["Manager", "Reflector", "Recorder", "Contributor"],
+  "canSubmit": ["Recorder"]
+}
+```
+
+| Attribute   | Type            | Default | Description                                                     |
+| ----------- | --------------- | ------- | --------------------------------------------------------------- |
+| `canView`   | Array of string | -       | The names of roles that can view this part of the assessment.   |
+| `canSubmit` | Array of string | -       | The names of roles that can submit this part of the assessment. |
+
+Setting either attribute to `[]` (empty array) means that **no role** can view/submit that part. If a question does not set the value of either attribute, it means that **every role** can view/submit that part.
```

Comment:
Nit: removing the word "question" since these attributes can be set (or not) at any level.

```suggestion
Setting either attribute to `[]` (empty array) means that **no role** can view/submit that part. If neither attribute is set, it means that **every role** can view/submit that part.
```

Reply:
Maybe this is worse, but my CS brain interpreted the original the wrong way (though it's obviously the wrong interpretation if you think about it for 2 seconds):

```suggestion
Setting either `canView` or `canSubmit` to `[]` (empty array) means that **no role** can view or submit that part, respectively. If a question does not set the value of either attribute, it means that **every role** can view/submit that part.
```

Reply:
(I'll leave it up to you to merge my two suggestions 🙂)

Reply:
```suggestion
Setting either `canView` or `canSubmit` to `[]` (empty array) means that **no role** can view or submit that part of the assessment, respectively. If either attribute is not set, it means that **every role** has the permission associated to the attribute, i.e., any student with any role can view or submit that part of the assessment.
```